src/Confluent.Kafka/Producer.cs

```diff
@@ -268,12 +282,17 @@ private void DeliveryReportCallbackImpl(IntPtr rk, IntPtr rkmessage, IntPtr opaq
         /// </summary>
         public int Poll(TimeSpan timeout)
         {
-            if (!manualPoll)
+            if (manualPoll)
             {
-                throw new InvalidOperationException("Poll method called, but manual polling is not enabled.");
+                return this.KafkaHandle.Poll((IntPtr)timeout.TotalMillisecondsAsInt());
             }
 
-            return this.KafkaHandle.Poll((IntPtr)timeout.TotalMillisecondsAsInt());
+            lock (pollSyncObj)
+            {
+                eventsServedCount = 0;
+                Monitor.Wait(pollSyncObj, timeout);
```

Comment:
I'm quite sure the only scenario your suggestion is more correct is if this method is called at the same time the poll loop has the `pollSyncObj` (which I also think is unlikely - scenarios in which this method is called probably correspond to librdkafka.poll blocking). Of course that's enough reason to change this, so change coming. Also, I think the logic of what you suggest is more straightforward to reason about than what I have - thanks.

Reply:
I believe that the semantics of this explicit Poll behaviour should be to return the number of events seen since the last explicit poll.

In your case you are returning the number of events seen since the explicit Poll() call was seen, but since the background thread is continously running Poll() it is likely that the waited-for event (DR) will already have been handled, in which case the explicit Poll() will block for the duration of the timeout, even though the expected event has already been seen.

Instead, if the semantics is for explicit Poll() to return the number of events since the last explicit Poll() there is no risk of blocking in the case the expected event has already been seen, since the explicit Poll() will return immediately when the count-since-last-Poll is > 0.

Reply:
Your Poll  with a queue size of 10.
```
    App thead              |   Background poll thread
   ------------------------+-------------------------
1.  Produce() OK x 10      |
2. Produce() QueueFull (queue limit reached) | 
3.                         |   rd_kafka_poll() serves 3 delivery reports, queue watermark is now below limit, sets eventPollCount += 3, trigger monitor which is a no-op since there is no listener.
4. Your explicit Poll(): set eventPollCount=0 |
5. Your explicit Poll(): blocks for 100ms |
6. Your explicit Poll(): returns 0.
```


My Poll with a queue size of 10.
```
    App thead              |   Background poll thread
   ------------------------+-------------------------
1.  Produce() OK x 10      |
2. Produce() QueueFull (queue limit reached) | 
3.                         |   rd_kafka_poll() serves 3 delivery reports, queue watermark is now below limit, sets eventPollCount += 3, trigger monitor which is a no-op since there is no listener.
4. My explicit Poll(): sees eventPollCount > 0 and returns it after setting to 0, no blocking |
```

Reply:
> Instead, if the semantics is for explicit Poll() to return the number of events since the last explicit Poll()

that's a key insight, will do that.